In [1]:
import numpy as np
import math

# Exercise 1

In [34]:
#defining velocities in the crustal
vs=3.6
vp=6.5


#defining the equations
#v=d/t, so vp=r/tp, vs=r/ts
#for ennouncement, first case: ts-tp=6 [s], second case ts-tp=8[s]

ts-tp=(r/vs)-(r/vp)
r=(ts-tp)*vp*vs/(vp-vs)

# Solve for the two possible locations for the earthquake. Give your results as the (x, y) km offset from station B

In [35]:
#first case:
ts1tp1=8
ra=np.around(((ts1tp1*vp*vs)/(vp-vs)),2)
print(f'ra= {ra} [Km]')

ra= 64.55 [Km]


In [36]:
#secondt case:
ts2tp2=6
rb=np.around((ts2tp2*vp*vs/(vp-vs)),2)
print(f'rb={rb} [Km]')

rb=48.41 [Km]


In [ ]:
#ra**2=(x-50)**2+y**2
#rb**2=x**2+y**2

In [37]:
from scipy.optimize import fsolve
def func(x):
    return [x[0]**2 + (x[1]-50)**2-ra**2,
            x[0]**2 + x[1]**2-rb**2
           ]


In [38]:
resultsroots=fsolve(func, [0, 0])
result0=np.round(resultsroots[0],3)
result1=np.round(resultsroots[1],3)
result1
print(f'The source locations are : x=+/- {result0}, y={result1}')

The source locations are : x=+/- 47.935, y=6.768


c) For each earthquake location, what is the predicted S − P time at Latteville?

In [39]:
#According to the announcement Lattevile has coordinates (100,25)
xlate=100
ylate=25
Late=[xlate,ylate]
Late

[100, 25]

In [40]:
#the sources locations are:
source1=[result0,result1]
source2=[-result0,result1]
print(f'According to our reference system source1 is onshore and source2 is offshore')
#We need to calculate the distance between Lateville and both sources.
r1=np.round(np.sqrt((Late[0]-source1[0])**2+(Late[1]-source1[1])**2),3)
print(f'The linear distance between Lateville and the source1 is {r1} [Km]')

According to our reference system source1 is onshore and source2 is offshore
The linear distance between Lateville and the source1 is 55.165 [Km]


In [41]:
r2=np.round(np.sqrt((Late[0]-source2[0])**2+(Late[1]-source2[1])**2),3)
print(f'The linear distance between Lateville and the source2 is {r2} [Km]')

The linear distance between Lateville and the source2 is 149.054 [Km]


In [49]:
#now we will calculate time differences.
tsource1late=np.round(r1*(vp-vs)/(vs*vp),3)

tsource2late=np.round(r2*(vp-vs)/(vs*vp),3)
print(f'the S − P time between lateville and source one (onshore) is {tsource1late} [s] and the S − P time between lateville and source two (offshore)is {tsource2late} [s]')

the S − P time between lateville and source one (onshore) is 6.837 [s] and the S − P time between lateville and source two (offshore)is 18.473 [s]


d) The strongest ground shaking usually results from S waves. Assuming the offshore location is correct, how much advance warning could be conveyed to Latteville before the S wave arrives at Latteville? Assume that the warning is possible as soon as the S wave has arrived at both stations.


In [54]:
timearrivals_1= np.round(ra/vs,3)#timearrival for station a
timearrivals_2=np.round(rb/vs,3)#timearrival for station b
timearrivals_3=np.round(r2/vs,3)#timearrival for station L-offshore
print(f'{timearrivals_3} [s], {timearrivals_2} [s], {timearrivals_1} [s]')

41.404 [s], 13.447 [s], 17.931 [s]


In [56]:
print(f'The time for warning is the difference between the last station and arrival time to Lateville {timearrivals_3-timearrivals_1}[s]')

The time for warning is the difference between the last station and arrival time to Lateville 23.473000000000003[s]


e) How much pre-S-wave warning could be provided to Latteville if the offshore earthquake location and origin time were known instantly (e.g., from a dense offshore network of seismometers)?

Ans: I think that if we have a dense network of seismometers offshore, we should have around 41.40 seconds to give the warning to Lateville.

# Exercise 2

Manipulation of moment tensors (chapter 9, question 2). This relates to chapter 9.2. An interesting seismic event occurred near Tori Shima, Japan, on June 13, 1984, which generated an unusually large tsunami for its size. Kanamori et al. (1993) argued that it was probably caused by magma injection and obtained the following components for the moment tensor: M11 = −1.8, M22 = −1.9, M33 = 3.7, M12 = −0.38, M31 = −0.96, M32 = 0.62 (all numbers in 1017 N m). To obtain stable results, they constrained the isotropic component to be zero. In this exercise, you will need to solve linear algebra. You may do this using scipy linalg toolbox (https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eig.html).

a)Compute M0 and MW for this event.

In [60]:
#according that we know to calculate M0, we need first built moment tensor MT

MT=(10**17)* np.array([[-1.8,-0.38,-0.96],[-0.38,-1.9,0.62],[-0.96,0.62,3.7]])
MT

array([[-1.8e+17, -3.8e+16, -9.6e+16],
       [-3.8e+16, -1.9e+17,  6.2e+16],
       [-9.6e+16,  6.2e+16,  3.7e+17]])

In [67]:
M0=(1/np.sqrt(2))*(np.sqrt(sum(MT.flatten()**2)))
MW=(2/3)*(np.log10(M0)-9.1)
print(f' The values for M0 and MW are {M0} and {MW} respectively')

 The values for M0 and MW are 3.42350697385006e+17 and 5.622980811255467 respectively


b)Compute the eigenvalues σ1, σ2, and σ3 (sorted such that σ1 > σ2 > σ3) and express M in its principal axes coordinates.

In [68]:
from numpy.linalg import eig

In [79]:
values,vectors = np.linalg.eig(MT)
print(f'The eigvalues of MT are {values}')
vvalues=np.array([values[0],values[2],values[1]])
vvalues
MTpaxes=np.array([[vvalues[0],0,0],
                 [0,vvalues[1],0],
                 [0,0,vvalues[2]]])
MTpaxes

The eigvalues of MT are [ 3.94087273e+17 -2.23977183e+17 -1.70110090e+17]


array([[ 3.94087273e+17,  0.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00, -1.70110090e+17,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00, -2.23977183e+17]])

c) Compute the parameter (equation 9.17), the measure of the misfit with a double-couple source. Is its value close to that expected for a pure double-couple (DC) source or a pure compensated linear vector dipole (CLVD) source? 

In [80]:
epsilon=np.round(vvalues[1]/abs(vvalues[0]),3)
epsilon

-0.432

d) Decompose M into MDC and MCLV D using Equation (9.14) and compute the scalar moment of each part.

In [82]:
MDC = np.array([[(1/2)*(vvalues[0]-vvalues[2]),0,0],
                [0,0,0],
                [0,0,-(1/2)*(vvalues[0]-vvalues[2])]
    
    
])
MDC
                

array([[ 3.09032228e+17,  0.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00, -3.09032228e+17]])

In [85]:
MCLVD=np.array([[-vvalues[1]/2,0,0],
                [0,vvalues[1],0],
                [0,0,-vvalues[1]/2]
    
    
])
MCLVD

array([[ 8.5055045e+16,  0.0000000e+00,  0.0000000e+00],
       [ 0.0000000e+00, -1.7011009e+17,  0.0000000e+00],
       [ 0.0000000e+00,  0.0000000e+00,  8.5055045e+16]])

In [87]:
#computing scalars numbers
MDC_scalar= (1/np.sqrt(2))*(np.sqrt(sum(MDC.flatten()**2)))
MDC_scalar
MCLVD_scalar = (1/np.sqrt(2))*(np.sqrt(sum(MCLVD.flatten()**2)))
MCLVD_scalar

1.4731965929674784e+17

e) Devise an alternative decomposition of M into MDC and MCLVD that maximizes the CLVD part and compute the scalar seismic

In [94]:
FRACCION=MDC_scalar/M0
FRACCION
print(f'{100*FRACCION} is % of fraction of total moment of MCLVD(compensate linear vector dipole)')

90.26773727307254 is % of fraction of total moment of MCLVD(compensate linear vector dipole)
